### DISCUSSION TITLE
Question about parameter's description in tools

### DISCUSSION THREAD

**Discussion by arielsch74** [RageAgainstThePixel/OpenAI-DotNet#392] [Q&A] [answered]
Using the provided example in the playground shows something like this:
```json
{
  "name": "get_weather",
  "description": "Determine weather in my location",
  "strict": true,
  "parameters": {
    "type": "object",
    "properties": {
      "location": {
        "type": "string",
        "description": "The city and state e.g. San Francisco, CA"
      },
      "unit": {
        "type": "string",
        "enum": [
          "c",
          "f"
        ]
      }
    },
    "additionalProperties": false,
    "required": [
      "location",
      "unit"
    ]
  }
}
```
where each parameter has a description.

When I use the OpenAI.Tool.GetOrCreateTool for a similar class it generates:
```json
{
  "name": "GetCurrentWeatherCTAsync_aab8da0fe2feb53b7aede73494ef5710",
  "description": "Cuando el usuario te consulte sobre el clima en algún lugar geográfico, debes pasar el nombre del lugar como parámetro.",
  "strict": false,
  "parameters": {
    "type": "object",
    "properties": {
      "city": {
        "type": "string"
      }
    },
    "required": [
      "city"
    ],
    "additionalProperties": false
  }
}
```
without desctiption of the parameter. 

My class is:
```csharp
namespace bizzuWeatherCustomTool;

public class WeatherCustomTool
{
/// <summary>
/// Asynchronously gets the current weather for the specified city.
/// </summary>
/// <param name="city">The name of the city to get the weather for.</param>
/// <returns>A task that represents the asynchronous operation. The task result contains a string with the current weather information or an error message.</returns>
public static async Task<string> GetCurrentWeatherCTAsync(string city)
    {
        try
        {
            return "{hot}"; 
        }
        catch (Exception ex)
        {
            return $"Error al obtener el clima: {ex.Message}";
        }
    }
}
```
Is it important for the model to have descriptions?

**Comment by StephenHodgson** [comment]
The descriptions just help the llm make a decision about what values to return and how to use the function call. They're optional, but help guide the model to make more accurate responses.

**Reply by arielsch74** [reply]
Thanks Setephen.

Is there a way in which I could provide the parameter's description using OpenAI.Tool.GetOrCreateTool method?

**Reply by StephenHodgson** [reply, marked as answer]
There are attributes for defining function calls: 
https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/ab908acd40bc458178d77fcd7765011cef9fc531/OpenAI-DotNet-Tests/TestServices/WeatherService.cs#L8-L32

- [`Function`](https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/main/OpenAI-DotNet/Common/FunctionAttribute.cs) to define the function and its description
- [`FunctionParamter`](https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/main/OpenAI-DotNet/Common/FunctionParameterAttribute.cs) to define the parameters and its description
- [`FunctionProperty`](https://github.com/RageAgainstThePixel/OpenAI-DotNet/blob/main/OpenAI-DotNet/Common/FunctionPropertyAttribute.cs) to define the properties and their default options

**Reply by arielsch74** [reply]
Thanks a lot!

**Reply by StephenHodgson** [reply]
Just keep in mind these attributes may get removed in the future as [.net 9 json schema](https://github.com/RageAgainstThePixel/OpenAI-DotNet/issues/377) is more powerful than my ad-hoc schema generator.

**Comment by arielsch74** [comment]
Hi Stephen!, I need to set the Strict property of the function parameter's definition to true
The Strict Property of the Tool's Function Property is readonly and the GetOrCreateTool Method doesn't have a parameter to set the strict property. Is there any way to do what I'm trying to do?

**Reply by StephenHodgson** [reply]
are you using json schema or just the property attributes?

**Reply by arielsch74** [reply]
The attributes for the Function and its parameters
```csharp
public class WeatherCustomTool
    {
    [Function("Get the current weather in a given location.")]
    public static async Task<string> GetCurrentWeatherCTAsync(
        [FunctionParameter("The location that the user wants to know how is the weather")] string location
        )
        {
            try
            {
            // var response = await _httpClient.GetStringAsync(apiUrl);
                return "{caluroso}"; // Devuelve la respuesta JSON con el clima actual (puedes procesarlo más)
            }
            catch (Exception ex)
            {
                return $"Error al obtener el clima: {ex.Message}";
            }
        }
    }
}
```
